### PR TITLE
Add create-remote-file-node-factory

### PR DESCRIPTION
**Pull request by teimurjan**
## Description

I've had a problem with [gatsby-source-prismic](https://github.com/angeloashmore/gatsby-source-prismic) during downloading a big amount of files. The problem was caused by `Request timeout error` raised inside the function `createRemoteFileNode` in *gatsby-source-filesystem*. After a day of debugging, I have found that the problem is in  [200 concurrent tasks](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-filesystem/src/create-remote-file-node.js#L102) in *gatsby-source-filesystem*. So I have created a new exported function called `createRemoteFileNodeFactory` which allow to choose your own options for maximum concurrent tasks and also for disabling progress logging(as during the downloading about 1000 files it's pretty noisy)

### CODE REVIEW

**Comment by pieh** [comment]
@teimurjan what's the concurrent task limit you used? Maybe we should just lower the default non configurable one?

**Comment by teimurjan** [comment]
@pieh, you know in different situations it’s needed to have different tasks number. In my case it was 5. I don’t know whether it was Prismic DDoS security or my laptop couldn’t run all well. So I think it would be be nice to have a way to control it by yourself

**Comment by pieh** [comment]
We have idea of making file downloader smarter - https://github.com/gatsbyjs/gatsby/issues/10049 so it could auto adjust concurrent connection limit (based on error rates), add auto-retry, etc. but this is not implemented yet.

I'm bit worried about adding this API like this, it would potentially limit us in the future (having to maintain this api, even if it doesn't really fit new design). But perhaps in this case if we finally implement the above idea, we could deprecate this and make this factory really only return passthrough function to new implementation (so no new download managers effectively would be used, but API would still be usable).

Will need to think a bit about this. But if we decide to move forward with factory approach, I think we would want to swap default `createRemoteFileNode` to use that factory (to avoid code duplication):

in `create-remote-file-node.js`
```js
const createRemoteFileNodeFactory = require(`./create-remote-file-node-factory`)
module.exports = createRemoteFileNodeFactory()
```

**Comment by teimurjan** [comment]
I added it with duplication as I didn't want to make a breaking change. So okay, will be waiting for your decision

**Comment by KyleAMathews** [comment]
@teimurjan this a super interesting approach!

It'd be nice to make things a bit more automatic though — we try hard to avoid config whenever possible.

What do you think about the idea of automatically scaling up and down the number of concurrent downloads?

**Comment by teimurjan** [comment]
It's a really good idea!

**Comment by LekoArts** [comment]
Hey @teimurjan 	👋 	Would be really cool to have this feature merged in, do you have time? 😊 	Also, @pieh & @KyleAMathews 	any further comments from you?

**Comment by teimurjan** [comment]
@LekoArts Hey! I've pushed it. Check it out 😁

File: packages/gatsby-source-filesystem/src/create-remote-file-node-factory.js [hunk 1]
```diff
@@ -0,0 +1,262 @@
+jest.mock(`fs-extra`, () => {
```
Comment (DSchau) [inline packages/gatsby-source-filesystem/src/create-remote-file-node-factory.js]: Little confused here--this should be in __tests__ right? e.g.

`packages/gatsby-source-filesystem/src/__test__/create-remote-file-node-factory.js`

File: packages/gatsby-source-filesystem/src/index.js [hunk 2]
```diff
@@ -5,6 +5,7 @@ function loadNodeContent(fileNode) {
 }
 
 exports.createFilePath = require(`./create-file-path`)
-exports.createRemoteFileNode = require(`./create-remote-file-node`)
+exports.createRemoteFileNodeFactory = require(`./create-remote-file-node-factory`)
+exports.createRemoteFileNode = exports.createRemoteFileNodeFactory()
```
Comment (DSchau) [inline packages/gatsby-source-filesystem/src/index.js]: Nit, but would probably prefer something like this:

```js
const createRemoteFileNodeFactory = require('`./create-remote-file-node-factory`)

// later

exports.createRemoteFileNodeFactory = createRemoteFileNodeFactory
exports.createRemoteFileNode = createRemoteFileNodeFactory()
```

**Comment by DSchau** [comment]
Going to close this per my comments [here](https://github.com/gatsbyjs/gatsby/pull/11788#pullrequestreview-244002449).

That being said, want to make two things super clear:

1. Thanks so much for this PR! We appreciate all the work you've done thus far.
2. We hope you'd be willing to build upon what was done here and try your hand at implementing #10049

Again -- thanks so much!

**Pull request closed without merging**